### PR TITLE
add local sgd example for resnet training

### PR DESCRIPTION
**Pull request by ForFishes**
1、add local sgd example for resnet training.
2、add warmup step for local sgd.
3、add warmup strategy for local sgd.

### CODE REVIEW

File: examples/local_sgd/resnet/train_gpu.sh [hunk 1]
```diff
@@ -15,12 +15,12 @@ MODEL_SAVE_PATH="output/"
...
 NUM_EPOCHS=90
-BATCH_SIZE=32
-LR=0.001
+BATCH_SIZE=128
+LR=0.1
 LR_STRATEGY=piecewise_decay
 
 # data params
-DATA_PATH="./ImageNet"
+DATA_PATH="/ssd2/shenliang/ImageNet"
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_gpu.sh]: donot use absolute path. And u should write some note to make user aware this data configuration

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 2]
```diff
@@ -47,9 +47,9 @@ def set_paddle_flags(flags):
 from paddle.fluid.contrib.mixed_precision.decorator import decorate
 import utils.reader_cv2 as reader
 from utils.utility import add_arguments, print_arguments, check_gpu
-from utils.learning_rate import cosine_decay_with_warmup
+from utils.learning_rate import cosine_decay_with_warmup, lr_warmup
 from paddle.fluid.contrib.mixed_precision.fp16_lists import black_list, white_list, gray_list
-from paddle.fluid.contrib.mixed_precision.fp16_utils import create_master_params_grads, master_param_to_train_param, rewrite_program
+#from paddle.fluid.contrib.mixed_precision.fp16_utils import create_master_params_grads, master_param_to_train_param, rewrite_program
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: donot keep comments of codes

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 3]
```diff
@@ -255,8 +282,8 @@ def build_program(is_train, main_prog, startup_prog, args, dist_strategy=None):
     image_shape = [int(m) for m in args.image_shape.split(",")]
     model_name = args.model
     model_list = [m for m in dir(models) if "__" not in m]
-    assert model_name in model_list, "{} is not in lists: {}".format(args.model,
-                                                                     model_list)
+ #   assert model_name in model_list, "{} is not in lists: {}".format(args.model,
+  #                                                                   model_list)
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: remote comments of code

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 4]
```diff
@@ -306,12 +333,13 @@ def build_program(is_train, main_prog, startup_prog, args, dist_strategy=None):
...
-                    optimizer.apply_gradients(master_params_grads)
-                    master_param_to_train_param(master_params_grads,
-                                                params_grads, main_prog)
+                 #   params_grads = optimizer.backward(avg_cost, startup_prog)
+                 #   master_params_grads = create_master_params_grads(
+                 #       params_grads, main_prog, startup_prog, args.scale_loss)
+                 #   optimizer.apply_gradients(master_params_grads)
+                 #   master_param_to_train_param(master_params_grads,
+                 #                               params_grads, main_prog)
+                     pass
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: remove comments of codes

File: examples/local_sgd/resnet/utils/reader_cv2.py [hunk 5]
```diff
@@ -47,8 +47,17 @@ def reader():
...
+                for passid in range(num_epoch):
+                    if shuffle:
+                        random.Random(pass_id_as_seed).shuffle(full_lines)
+
+                    temp_lines = list(full_lines[trainer_id * per_node_lines:(trainer_id + 1)
+                                            * per_node_lines])
+                    lines.extend(temp_lines)
+                random.Random(pass_id_as_seed).shuffle(lines) 
+               # lines = full_lines[trainer_id * per_node_lines:(trainer_id + 1)
+                #                   * per_node_lines]
```
Comment (gavin1332) [inline examples/local_sgd/resnet/utils/reader_cv2.py]: remote comments of code

File: examples/local_sgd/resnet/utils/reader_cv2.py [hunk 6]
```diff
@@ -32,7 +32,7 @@ def _reader_creator(settings,
                     color_jitter=False,
                     rotate=False,
                     data_dir=DATA_DIR,
-                    pass_id_as_seed=0):
+                    pass_id_as_seed=0,num_epoch=0):
```
Comment (gavin1332) [inline examples/local_sgd/resnet/utils/reader_cv2.py]: illegal format

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 7]
```diff
@@ -434,115 +484,157 @@ def if_exist(var):
...
+    test_acc1 = np.array(test_info[1]).mean()
+    test_acc5 = np.array(test_info[2]).mean()
+
+    print("test_loss {0}, test_acc1 {1}, test_acc5 {2}".format("%.5f"%test_loss,\
+          "%.5f"%test_acc1, "%.5f"%test_acc5))
+    sys.stdout.flush()
+    model_path = os.path.join(model_save_dir + '/' + model_name, str(pass_id))
+    if not os.path.isdir(model_path):
+        os.makedirs(model_path)
+    fluid.io.save_persistables(exe, model_path, main_program=fleet._origin_program)
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: have u ever tested warm-start?

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 8]
```diff
@@ -434,115 +484,157 @@ def if_exist(var):
...
+                      pass_id, "%.5f"%train_loss, "%.5f"%train_acc1, "%.5f"%train_acc5, "%.2f" % train_speed))
+                sys.stdout.flush()
+                #init 
+                batch_id = 0
+                train_info = [[], [], []]
+                train_begin=time.time()
+                batch_id = 0
+                time_record=[]
+
+                if args.isTest:#test
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: what's #test for? i think the if-condition is clear enough

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 9]
```diff
@@ -434,115 +484,157 @@ def if_exist(var):
...
+                    test_info = [[], [], []]
+                    test_py_reader.start()
+                    test_batch_id = 0
+                    try:
+                        while True:
+                            t1 = time.time()
+                            loss, acc1, acc5 = exe.run(program=test_prog,\
+                                                       fetch_list=test_fetch_list)
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: missing parameter `use_program_cache=True`, which speed up prediction performance a lot.

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 10]
```diff
@@ -434,115 +484,157 @@ def if_exist(var):
...
-    for pass_id in range(params["num_epochs"]):
+
+    global_batch_size = args.batch_size * num_trainers
+    steps_per_pass = int(math.ceil(args.total_images * 1.0 / global_batch_size))
+    print("steps_per_pass  {}".format(steps_per_pass))
+    print("global_batch_size {}".format(global_batch_size))
+
+    pass_id = 0
+    # while 
+   # train_py_reader.start()
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: remove code comments

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 11]
```diff
@@ -434,115 +484,157 @@ def if_exist(var):
...
+
+    #start test
+    test_py_reader.start()
+    test_batch_id = 0
+    test_info = [[], [], []]
+    try:
+        while True:
+            t1 = time.time()
+            loss, acc1, acc5 = exe.run(program=test_prog,\
+                                       fetch_list=test_fetch_list)
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: u should better help creating a distributed prediction algorithm in the furture.

File: examples/local_sgd/resnet/train_gpu.sh (hunk 1 above)
Comment (gavin1332) [reply examples/local_sgd/resnet/train_gpu.sh]: test

Comment (ForFishes) [reply examples/local_sgd/resnet/train_gpu.sh]: test2

File: examples/local_sgd/resnet/train_gpu.sh [hunk 12]
```diff
@@ -29,7 +30,7 @@ IMAGE_SHAPE=3,224,224
 #gpu params
 FUSE=True
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_gpu.sh]: not used

File: examples/local_sgd/resnet/train_gpu.sh [hunk 13]
```diff
@@ -29,7 +30,7 @@ IMAGE_SHAPE=3,224,224
...
 NCCL_COMM_NUM=2
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_gpu.sh]: have u ever check that this configuration is applied?

File: examples/local_sgd/resnet/train_gpu.sh [hunk 14]
```diff
@@ -29,7 +30,7 @@ IMAGE_SHAPE=3,224,224
...
-NUM_CARDS=4
+NUM_CARDS=8
 FP16=False #whether to use float16
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_gpu.sh]: is this configuration applicable? in my mind, local sgd have not been tested in FP16 mode

Comment (gavin1332) [reply examples/local_sgd/resnet/train_gpu.sh]: If it is not applicable, u'd better remove it.

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 15]
```diff
@@ -89,30 +88,56 @@ def set_paddle_flags(flags):
...
 add_arg('use_gpu',                      bool,   True,                 "Whether to use GPU or not.")
 add_arg('nccl_comm_num',                int,    1,                    "nccl comm num")
-add_arg('num_iteration_per_drop_scope', int,    10,                   "Ihe iteration intervals to clean up temporary variables.")
+add_arg('num_iteration_per_drop_scope', int,    30,                   "Ihe iteration intervals to clean up temporary variables.")
 
 add_arg('use_local_sgd',                bool,   True,                 "Whether to use LocalSGD argorithmn.")
 add_arg('local_sgd_steps',              int,    2,                    "The step number for local training before synchronizing parameters.")
+add_arg('local_sgd_is_warm_steps',      int,    30,                   "The warmup step of number for local sgd.")
+add_arg('lsgd_warmup_strategy',         int,    1,                    "Select strategy to warmup the lsgd,1:exp,2:const,3:linear")
+add_arg('isTest',                       bool,   False,                "Whether to test on every epoch")
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: why the format of this argument is different to others, you should prefer `is_test`.

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 16]
```diff
@@ -89,30 +88,56 @@ def set_paddle_flags(flags):
...
+
+
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: too many blank lines.
reference: http://google.github.io/styleguide/pyguide.html#35-blank-lines

File: examples/local_sgd/resnet/train_with_fleet.py [hunk 17]
```diff
@@ -89,30 +88,56 @@ def set_paddle_flags(flags):
...
 
 def optimizer_setting(params):
     ls = params["learning_strategy"]
     l2_decay = params["l2_decay"]
     momentum_rate = params["momentum_rate"]
     if ls["name"] == "piecewise_decay":
```
Comment (gavin1332) [inline examples/local_sgd/resnet/train_with_fleet.py]: are u sure that these branches of `ls["name"]` == xxx works well? if u are not sure, just keep what u have tested

Comment (gavin1332) [reply examples/local_sgd/resnet/train_with_fleet.py]: It's a hard work. We can postpone these test.

File: examples/local_sgd/resnet/train_gpu.sh (hunk 14 above)
Comment (ForFishes) [reply examples/local_sgd/resnet/train_gpu.sh]: Done

File: examples/local_sgd/resnet/train_with_fleet.py (hunk 15 above)
Comment (ForFishes) [reply examples/local_sgd/resnet/train_with_fleet.py]: Done

File: examples/local_sgd/resnet/train_with_fleet.py (hunk 16 above)
Comment (ForFishes) [reply examples/local_sgd/resnet/train_with_fleet.py]: Done

File: examples/local_sgd/resnet/train_with_fleet.py (hunk 17 above)
Comment (ForFishes) [reply examples/local_sgd/resnet/train_with_fleet.py]: remove these..
